### DISCUSSION TITLE
17032023 all rules and first bug fixes

### DISCUSSION THREAD

**Discussion by tab1tha** [data-to-insight/CIN-validator#405] [Announcements]
## What's Changed
* XML to DataFrames by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/113
* [xml-ingress] ChildIdentifiersTable, ChildCharacteristicsTable, CINdetailsTable by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/114
* black formatting by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/115
* Rule template: Demonstrate on Rule 8500 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/117
* setup devcontainer to install depenency packages by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/119
* Rule 8510, created and working by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/118
* add pre-commit to requirements by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/121
* Rule100 completed by @JohnCWaC in https://github.com/data-to-insight/CIN-validator/pull/125
* Added Rule 1510 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/124
* Rule1530 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/126
* Makes census period from collection_year by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/123
* Established rule1550 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/127
* Added4220 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/128
* Added8790 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/129
* Rule1540 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/130
* Added8794 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/131
* Added8650 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/133
* Ops/GitHub actions by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/134
* Rule 4180 by @amyrichards0n in https://github.com/data-to-insight/CIN-validator/pull/135
* Added rule_8617 #72 by @bellt101 in https://github.com/data-to-insight/CIN-validator/pull/137
* Added8568 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/140
* define query RuleType by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/142
* rule-8870q by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/141
* Generate tables from XML by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/145
* Run all rules on data files by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/146
* Type1 rule implementation by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/150
* Added8545Q by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/143
* Added8905 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/148
* Rule_8840 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/152
* use ReferenceDate in make_census_period by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/153
* Rule8715 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/147
* Rule8930 Closes #106 by @amyrichards0n in https://github.com/data-to-insight/CIN-validator/pull/154
* Closes Rule #65 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/151
* Added 8535Q by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/155
* No errors closes 8520 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/156
* Added8615 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/160
* Added rule 8608 by @WendyWigan in https://github.com/data-to-insight/CIN-validator/pull/159
* Revised4012Q by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/157
* Rule structure: Multiple columns across multiple tables. by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/166
* Working rule8614 by @Csmyth201 in https://github.com/data-to-insight/CIN-validator/pull/164
* XMLtoCSV CLI command by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/161
* Rule8805 by @Csmyth201 in https://github.com/data-to-insight/CIN-validator/pull/163
* Added rule 4011 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/158
* Added8630 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/170
* Edit rule 8630 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/174
* Rule_8640 complete closes#43 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/177
* Added rule 1105 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/138
* Created rule 8915 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/181
* Implement type3 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/182
* Rule 8896: example of type3 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/172
* added 8935 - will's type 3 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/175
* Rule8525 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/176
* Rule8810 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/168
* Rule8841 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/171
* Rule8897 by @WendyWigan in https://github.com/data-to-insight/CIN-validator/pull/173
* Rule8730 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/178
* Proof of concept: type3 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/185
* 8736 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/179
* Fixes print statements, adds date input flexibility for reference date by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/183
* fixed make_date fucntion to allow varied inputs by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/188
* Rule_4004 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/187
* Rule 2885: Example of type2. by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/167
* Fix group keys by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/191
* Feature/check per type by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/194
* Feature/error df by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/199
* Added rule 8750 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/169
* rule 4008 working, formatted by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/190
* Rule8920 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/192
* rule 8555 created by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/193
* Created rule 8585Q by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/195
* Rule1103 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/197
* Created rule 8875 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/196
* Rule_8565 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/200
* group keys set to false in rules raising warnings by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/202
* Rule4013 by @Csmyth201 in https://github.com/data-to-insight/CIN-validator/pull/201
* cli command to select issues by ERROR_IDs by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/210
* Added4015 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/184
* fixes group keys in 4008 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/212
* Added rule 8740 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/189
* Added rule 4001 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/208
* Resolve erratic behaviour in run-all command by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/214
* Include run instructions in README.MD by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/215
* Isolate functions from CLI by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/216
* Rule8620 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/206
* rule 8530Q by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/204
* Added rule 8940 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/209
* Rule8867 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/205
* Rule8720 Issue #101 by @amyrichards0n in https://github.com/data-to-insight/CIN-validator/pull/198
* Rule 8670Q by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/203
* Added rule 4016 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/207
* draft Rules df by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/217
* Added rule 4017 (mirror of 4016) by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/219
* Added rule 8839 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/220
* Create RULE_PATTERNS.MD by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/230
* Rule 8606 by @phoebefryer in https://github.com/data-to-insight/CIN-validator/pull/227
* Added rule 8590 by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/221
* 8842Q by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/223
* rule 8832 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/224
* 4000 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/225
* Rule 1104 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/228
* Rule8675Q  by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/226
* 8869 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/232
* Added rule 8770Q by @SLornieCYC in https://github.com/data-to-insight/CIN-validator/pull/241
* 8873 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/233
* Rule 4010v2 closes #79 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/245
* 8899Q working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/234
* 4003 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/235
* AddsRule4014 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/231
* 2990 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/238
* Feature/generate output by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/236
* 8890 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/239
* 8820 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/240
* 8815 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/242
* 8831 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/243
* 8825Q working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/244
* Rule8600 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/248
* 8775Q working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/246
* 8868 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/247
* 8540 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/250
* 8569 working by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/252
* Implement return level validation by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/256
* Rule 2886Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/257
* Rule 1520 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/258
* Rule2887Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/259
* Correct file type: 2886Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/260
* Rule2883 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/262
* edits on 8866 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/263
* 8866 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/253
* 8610 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/255
* Added2889 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/136
* Rule 2888Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/265
* 2884 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/254
* 4009 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/266
* Code for review Closes Rule #69 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/211
* 8816 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/251
* 8898 working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/267
* update comments: 8898 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/268
* Rule8772 by @GBurrel in https://github.com/data-to-insight/CIN-validator/pull/261
* 2991Q working by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/264
* Rule 1560Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/269
* Added8910 by @PhilDeakin in https://github.com/data-to-insight/CIN-validator/pull/149
* Feature/rpc layer by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/270
* Process date columns by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/277
* Process RemoteFile objects by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/278
* parse RemoteFile by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/279
* Refactor/integration fixes - Json Serialiser by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/281
* Refactor/rule sweep by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/229
* Process date columns - more efficient implimentation by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/282
* Include child ID in issue report by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/285
* typo correction: "sources" not "sourcing" by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/286
* Documentation by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/237
* Docstrings for all central functions by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/280
* feature/select rules by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/289
* Rule sweep top down 100-8614 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/287
* Contributor docs by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/290
* Fix dateofinitialcpc bug by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/293
* Refactor/rule sweep ascend by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/288
* Fix 8816 - ValueError: attempt to get argmax of an empty sequence by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/295
* diagram.data flow in cin tool. by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/296
* debug/issue_report: row_id as integers by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/298
* Refactor/remove data container by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/299
* Feature/include bank holidays by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/300
* Fix/xml csv missing data by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/301
* separate data processing from conversion by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/302
* enable successful ingress with incomplete data by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/303
* Fix/no error scenario by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/304
* fix lint error. isort by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/308
* Feature/create report by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/306
* added simple timer to CLI for testing purposes by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/307
* Bugfix 8535Q by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/327
* remove hardcoded cin_data by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/332
* 1560Q now correctly checks <FormerUPN> not <UPN> by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/325
* 4180 bugfix by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/326
* 8832 bugfix by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/328
* bugfix 8897  by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/329
* bugfix 8869 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/330
* bugfix 8868 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/333
* 8620 bugfix by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/334
* removes select-issue-by-id feature by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/335
* Bugfix 8863 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/338
* Bugfix 8820 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/340
* Bugfix 8839 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/339
* bugfix 4001, added check for CINplanStartDate by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/337
* bugfix 8890 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/336
* Docs/typos by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/349
* Bugfix 2885 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/341
* Bugfix/fix inequalities by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/351
* 2324ruleset/errors to queries by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/347
* Bugfix/fix merges by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/352
* bugfix 4017  by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/354
* bugfix 1510 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/356
* fixed 8535Q by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/357
* bugfix 1560Q by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/358
* remove TODOs by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/350
* 1560Q fix pytest by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/359
* Fix/key error 4001 by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/360
* mvp init file by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/361
* suppress false-positive SettingWithCopyError by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/362
* ensure dates are passed as strings to fe by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/363
* apply datetime_to_str recursively by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/364
* bugfix 4014 to flag duplicate cin plans by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/386
* Bugfix 8816 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/387
* Bugfix/8839 by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/389
* bugfix 8675Q, by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/392
* Bugfix 2888Q by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/394
* bugfix 8820 correctly flags duplicates by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/388
* bugfix 2991Q  by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/395
* bugfix 8670Q by @willLPD2I in https://github.com/data-to-insight/CIN-validator/pull/391
* replace truthy check with explicit len check by @tab1tha in https://github.com/data-to-insight/CIN-validator/pull/401

## New Contributors
* @willLPD2I made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/119
* @phoebefryer made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/118
* @JohnCWaC made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/125
* @PhilDeakin made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/124
* @amyrichards0n made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/135
* @bellt101 made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/137
* @GBurrel made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/147
* @WendyWigan made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/159
* @Csmyth201 made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/164
* @SLornieCYC made their first contribution in https://github.com/data-to-insight/CIN-validator/pull/169

**Full Changelog**: https://github.com/data-to-insight/CIN-validator/commits/0.1.1

<hr /><em>This discussion was created from the release <a href='https://github.com/data-to-insight/CIN-validator/releases/tag/v0.1.1'>17032023 all rules and first bug fixes</a>.</em>

**Comment by tab1tha** [comment]
@SLornieCYC a new version of the tool is live.
If you get a chance, could you check whether the bug fixes worked?

@willLPD2I

**Reply by WillLP-code** [reply]
note: this isn't everything you raised Stephen, if somethin is still in issues we haven't made a PR to deal with it yet

**Reply by SLornieCYC** [reply]
@tab1tha The implemented bug fixes all look to be working as expected from a quick scan through (comparing pre- and post-fix CIN validator error outputs and then matching to DFE portal output as well).  

The only one I'm not 100% sure about is the working days part on 8670Q. https://github.com/data-to-insight/CIN-validator/issues/377#issuecomment-1476100296

**Comment by tab1tha** [comment]
@SLornieCYC @willLPD2I  a new version of the tool is now live. 
If you get a chance, do check whether it works well now.

Happy Easter !
